### PR TITLE
m4: add 1.4.19 + fix compilation with clang on ubuntu

### PR DESCRIPTION
**Pull request by madebr**
Specify library name and version:  **m4/all**

- for 1.4.18: fix building with clang on Ubuntu (it failed building with e.g. a stock clang10 from ubuntu 20.04)
- add 1.4.19: ported only the `src` patch, I assume all the patches to the `lib` subfolder are already fixed by gnulib

---

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/m4/all/conanfile.py [hunk 1]
```diff
@@ -98,13 +99,15 @@ def build(self):
...
     def package(self):
-        self.copy(pattern="COPYING", src=self._source_subfolder, dst="licenses")
+        self.copy("COPYING", src=self._source_subfolder, dst="licenses")
         with self._build_context():
             autotools = self._configure_autotools()
             autotools.install()
         tools.rmdir(os.path.join(self.package_folder, "share"))
 
     def package_info(self):
+        self.cpp_info.libs = []
```
Comment (madebr) [inline recipes/m4/all/conanfile.py:109]:
```suggestion
        self.cpp_info.libdirs = []
```

**Comment by madebr** [comment]
Building m4 on Visual Studio 2019 with debug runtime (MDd/MTd) succeeds, but the test package fails.
I managed to create a simple reproducer that has the same behavior:
```c
#include <io.h>
#include <stdio.h>
int main() {
    int no =  _fileno (stdin);
    _close(no);
    fclose(stdin);
}
```
(compile with `cl /MDd repro.c /Ferepro.exe`
I can't find a configure option to avoid the double close.
This happens in `lib/fclose.c`.

I don't know why the test succeeds on MSV2017. Perhaps its debug runtime is less strict?

**Comment by SSE4** [comment]
what is the fix for clang in 1.4.19?

**Comment by madebr** [comment]
> what is the fix for clang in 1.4.19?

The missing symbols came from a function re-implemented by gnulib.
They updated the gnulib sources to this year's master, and apparently they fixed it.
I tested it with rhw ubuntu:20.04 image, using its clang10 compiler.

**Comment by uilianries** [comment]
Interesting, it should be automatically solved by the compiler. The new docker images are using compiler-rt by default with clang, so it should be "hidden".

**Review by uilianries — APPROVED** [review]

**Comment by madebr** [comment]
> Interesting, it should be automatically solved by the compiler. The new docker images are using compiler-rt by default with clang, so it should be "hidden".

The Conan docker images work fine.

**Review by prince-chrismc — APPROVED** [review]

**Comment by SSE4** [comment]
> @SSE4
> 
> Building m4 on Visual Studio 2019 with debug runtime (MDd/MTd) succeeds, but the test package fails (even `m4 --help` fails).

why doesn't it fail on CI?

**Comment by madebr** [comment]
> 
> 
> > @SSE4
> > Building m4 on Visual Studio 2019 with debug runtime (MDd/MTd) succeeds, but the test package fails (even `m4 --help` fails).
> 
> why doesn't it fail on CI?

CI built m4 with Visual Studio 2017.
Perhaps some difference there?

**Comment by SSE4** [comment]
> CI built m4 with Visual Studio 2017.
> Perhaps some difference there?

I can reproduce an issue locally in VS 2019. don't have VS 2017 at the moment to check.
that I can see:
[fclose](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/fclose-fcloseall?view=msvc-160)
```
if stream is NULL, the invalid parameter handler is invoked, as described in Parameter Validation.
```
[parameter validation](https://docs.microsoft.com/en-us/cpp/c-runtime-library/parameter-validation?view=msvc-160)
```
In debug builds, the invalid parameter macro usually raises a failed assertion and a debugger breakpoint before the dispatch function is called. When the code is executed, the assertion may be reported to the user in a dialog box that has "Abort", "Retry", and "Continue" or similar choices depending on the operating system and runtime library version.
```

I see in m4 code they install some custom invalid parameter handler (`HAVE_MSVC_INVALID_PARAMETER_HANDLER`), but it doesn't seem to help...

**Comment by madebr** [comment]
@SSE4 
I think we will need to post a message to the [gnulib mailing list](https://lists.gnu.org/mailman/listinfo/bug-gnulib).
Mentioning [this little reproducer](https://github.com/conan-io/conan-center-index/pull/7369#issuecomment-924508173) and that we are building with MSVC2019 using debug runtime (MDd/MTd).

**Comment by SSE4** [comment]
> @SSE4
> I think we will need to post a message to the [gnulib mailing list](https://lists.gnu.org/mailman/listinfo/bug-gnulib).
> Mentioning [this little reproducer](https://github.com/conan-io/conan-center-index/pull/7369#issuecomment-924508173) and that we are building with MSVC2019 using debug runtime (MDd/MTd).

I see that message is thrown from this line (`close.cpp` from UCRT):
```
 _VALIDATE_CLEAR_OSSERR_RETURN((_osfile(fh) & FOPEN), EBADF, -1);
```
which basically goes into:
```
#define _VALIDATE_CLEAR_OSSERR_RETURN( expr, errorcode, retexpr )              \
    {                                                                          \
        int _Expr_val=!!(expr);                                                \
        _ASSERT_EXPR( ( _Expr_val ), _CRT_WIDE(#expr) );                       \
        if ( !( _Expr_val ) )                                                  \
        {                                                                      \
            _doserrno = 0L;                                                    \
            errno = errorcode;                                                 \
            _INVALID_PARAMETER(_CRT_WIDE(#expr) );                             \
            return ( retexpr );                                                \
        }                                                                      \
    }

```
then `_ASSERT_EXPR` expands into:
```
#define _ASSERT_EXPR(expr, msg) \
        (void) ((!!(expr)) || \
                (1 != _CrtDbgReportW(_CRT_ASSERT, _CRT_WIDE(__FILE__), __LINE__, NULL, msg)) || \
                (_CrtDbgBreak(), 0))
```

so it's not `_INVALID_PARAMETER` who shows a message box, but `_ASSERT_EXPR`.
it uses this check (`dbgrprtt.cpp`, function `_VCrtDbgReport`):
```
if (_CrtDbgMode[nRptType] & _CRTDBG_MODE_WNDW)
```
the behavior of the function [_CrtDbgReportW](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/crtdbgreport-crtdbgreportw?view=msvc-160) is controlled by the [_CrtSetReportMode](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/crtsetreportmode?view=msvc-160). unfortunately I don't see it used in `m4`/`gnulib`.
I think that would be the easiest fix for m4 would be to call `_CrtSetReportMode` somewhere.
you can try adding to your repro:
```
_CrtSetReportMode(_CRT_ASSERT, 0);
```
maybe there is an easier fix with the compiler option (as debug builds emits it), we might need to check what exactly triggers that behavior (maybe `_DEBUG` or `DEBUG` macro).

**Review by SSE4 — APPROVED** [review]

**Pull request merged**